### ISSUE TITLE
Odd behavior with repeated reads

### ISSUE THREAD

**Issue by williamberman** [greghendershott/racket-mode#449]
I enter the following in a racket mode buffer

```
#lang racket

(provide the-repl)

(define (the-repl)
  (define (the-eval input)
    (format "received: ~a" input))

  (define (the-loop)
    (display "my-repl> ")
    (define the-input (read))
    (if (eq? the-input 'quit)
        (displayln "Goodbye")
        (begin
          (displayln (the-eval the-input))
          (the-loop))))

  (the-loop))
```

I run `racket-run`

```
repl.rkt> (the-repl)
my-repl> 1
received: 1
my-repl> 2
3
received: 3
my-repl> 4
received: 4
my-repl> quit
Goodbye
2
repl.rkt> 
```
When `2` is read, nothing happens. It seems like it's only read and displayed after quit is entered?

The function works as expected when run from a racket repl launched from the command line.

Let me know if there's anything I can do to further clarify. Thank you for your help.

To help me help you, please:

1. Open any racket-mode or racket-repl-mode buffer.

2. <kbd>M-x</kbd> `racket-bug-report` <kbd>ENTER</kbd>.

3. Copy and paste that, here.

<details>
```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
		 (20200407 1408)
		 "Major mode for Racket language."
		 ((emacs
		   (25 1))
		  (faceup
		   (0 0 2))
		  (pos-tip
		   (20191127 1028)))
		 nil nil "/Users/wlbberman/.emacs.d/elpa/racket-mode-20200407.1408"
		 ((:url . "https://www.racket-mode.com/")
		  (:maintainer "Greg Hendershott")
		  (:authors
		   ("Greg Hendershott")))
		 nil))
((emacs-version "26.1")
 (emacs-uptime "1 day, 2 hours, 57 minutes, 50 seconds")
 (system-type darwin)
 (major-mode racket-mode)
 (racket--el-source-dir "/Users/wlbberman/.emacs.d/elpa/racket-mode-20200407.1408/")
 (racket--rkt-source-dir "/Users/wlbberman/.emacs.d/elpa/racket-mode-20200407.1408/racket/")
 (racket-program "racket")
 (racket-command-port 55555)
 (racket-command-timeout 10)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-use-repl-submit-predicate nil)
 (racket-images-system-viewer "open")
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal))))
(enabled-minor-modes
 (async-bytecomp-package-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (blink-cursor-mode)
 (company-mode)
 (counsel-mode)
 (cursor-sensor-mode)
 (delete-selection-mode)
 (diff-auto-refine-mode)
 (display-line-numbers-mode)
 (electric-indent-mode)
 (file-name-shadow-mode)
 (flycheck-mode)
 (font-lock-mode)
 (global-auto-revert-mode)
 (global-company-mode)
 (global-display-line-numbers-mode)
 (global-eldoc-mode)
 (global-flycheck-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (global-magit-file-mode)
 (hs-minor-mode)
 (line-number-mode)
 (magit-auto-revert-mode)
 (menu-bar-mode)
 (mouse-wheel-mode)
 (racket-xp-mode)
 (rainbow-delimiters-mode)
 (shell-dirtrack-mode)
 (show-smartparens-global-mode)
 (show-smartparens-mode)
 (smartparens-global-strict-mode)
 (smartparens-mode)
 (smartparens-strict-mode)
 (tool-bar-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (yas-global-mode)
 (yas-minor-mode))
(disabled-minor-modes
 (abbrev-mode)
 (auto-fill-function)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (buffer-read-only)
 (cl-old-struct-compat-mode)
 (column-number-mode)
 (company-search-mode)
 (compilation-in-progress)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (cursor-intangible-mode)
 (defining-kbd-macro)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (display-time-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-pair-mode)
 (electric-quote-mode)
 (flymake-mode)
 (git-commit-mode)
 (global-prettify-symbols-mode)
 (global-visual-line-mode)
 (haskell-indentation-mode)
 (horizontal-scroll-bar-mode)
 (ido-everywhere)
 (isearch-mode)
 (ivy-mode)
 (jit-lock-debug-mode)
 (magit-blame-mode)
 (magit-blame-read-only-mode)
 (magit-blob-mode)
 (magit-file-mode)
 (magit-wip-after-apply-mode)
 (magit-wip-after-save-local-mode)
 (magit-wip-after-save-mode)
 (magit-wip-before-change-mode)
 (magit-wip-initial-backup-mode)
 (magit-wip-mode)
 (mail-abbrevs-mode)
 (mc-hide-unmatched-lines-mode)
 (mml-mode)
 (multiple-cursors-mode)
 (next-error-follow-minor-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (prettify-symbols-mode)
 (racket-smart-open-bracket-mode)
 (rectangle-mark-mode)
 (rectangular-region-mode)
 (server-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (size-indication-mode)
 (smartparens-global-mode)
 (smerge-mode)
 (temp-buffer-resize-mode)
 (transient-resume-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (which-function-mode)
 (window-divider-mode)
 (with-editor-mode)
 (xref-etags-mode))
```
</details>

**Comment by greghendershott** [comment]
Thank you for reporting this. I can reproduce.

This arose while handling #311. I think I've figured out a better way to handle this, which is pushed to a topic branch as commit 50d29b1.

I'd like to test and think about that a bit longer, before merging to `master`.

**Comment by sorawee** [comment]
Doesn't this fix #353 as well?
